### ISSUE TITLE
twisted.conch: publickey auth not working against gerrit/Jsch server

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5241]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| toidinamai reported|
|-|-|
|Trac ID|trac#5241|
|Type|defect|
|Created|2011-08-25 19:54:19Z|
|Branch|https://github.com/twisted/twisted/tree/conch-publickey-bool-5241|

Hi,

I'm trying to use conch to connect to an in-company gerrit instance to process
events and control changes.  Unfortunately conch fails when trying to connect
to gerrit, even though openssh works using the same key.  What is even weirder
is that I can use the same key to connect to my local openssh server and it works
with both conch and openssh.

I tried limiting the algorithms to the same that openssh uses (aes128-cbc
hmac-md5) but that didn't help either.

I'll try to do some more debugging and give more details as soon as possible
but right now the easiest way to reproduce is:

1. Set up account on https://review.source.android.com/
2. Register SSH key.
3. Connect to SSH server on port 29418 and you should see:

  ****    Welcome to Gerrit Code Review    ****

  Hi XXX, you have successfully connected over SSH.

  Unfortunately, interactive shells are disabled.
  To clone a hosted Git repository, use:

  git clone ssh://XXX@review.source.android.com:29418/REPOSITORY_NAME.git

4. Try same with conch:
2011-08-25 21:44:29+0200 [-] Log opened.
2011-08-25 21:44:29+0200 [-] Starting factory <twisted.conch.client.direct.SSHClientFactory instance at 0x244f320>
2011-08-25 21:44:35+0200 [SSHClientTransport,client] kex alg, key alg: diffie-hellman-group1-sha1 ssh-rsa
2011-08-25 21:44:35+0200 [SSHClientTransport,client] outgoing: aes128-cbc hmac-md5 none
2011-08-25 21:44:35+0200 [SSHClientTransport,client] incoming: aes128-cbc hmac-md5 none
2011-08-25 21:44:36+0200 [SSHClientTransport,client] REVERSE
2011-08-25 21:44:37+0200 [SSHClientTransport,client] NEW KEYS
2011-08-25 21:44:37+0200 [SSHClientTransport,client] setting client server to <twisted.conch.client.default.SSHUserAuthClient instance at 0x244f290>
2011-08-25 21:44:37+0200 [SSHClientTransport,client] starting service ssh-userauth
2011-08-25 21:44:37+0200 [SSHClientTransport,client] using agent
2011-08-25 21:44:37+0200 [SSHAgentClient,client] got 1 public keys
2011-08-25 21:44:38+0200 [SSHService ssh-userauth on SSHClientTransport,client] can continue with: ['publickey']
2011-08-25 21:44:38+0200 [SSHService ssh-userauth on SSHClientTransport,client] trying to auth with publickey
2011-08-25 21:44:38+0200 [SSHService ssh-userauth on SSHClientTransport,client] using key of type RSA
2011-08-25 21:44:39+0200 [SSHService ssh-userauth on SSHClientTransport,client] can continue with: ['publickey']
2011-08-25 21:44:39+0200 [SSHService ssh-userauth on SSHClientTransport,client] trying to auth with publickey
2011-08-25 21:44:39+0200 [SSHService ssh-userauth on SSHClientTransport,client] ['review.source.android.com.key']
2011-08-25 21:44:39+0200 [SSHService ssh-userauth on SSHClientTransport,client] ['review.source.android.com.key']
2011-08-25 21:44:39+0200 [SSHService ssh-userauth on SSHClientTransport,client] review.source.android.com.key
2011-08-25 21:44:39+0200 [SSHService ssh-userauth on SSHClientTransport,client] using key of type RSA
2011-08-25 21:44:42+0200 [SSHService ssh-userauth on SSHClientTransport,client] can continue with: ['publickey']
2011-08-25 21:44:42+0200 [SSHService ssh-userauth on SSHClientTransport,client] trying to auth with publickey
2011-08-25 21:44:42+0200 [SSHService ssh-userauth on SSHClientTransport,client] ['review.source.android.com.key']
2011-08-25 21:44:42+0200 [SSHService ssh-userauth on SSHClientTransport,client] []
2011-08-25 21:44:42+0200 [SSHService ssh-userauth on SSHClientTransport,client] Disconnecting with error, code 14
        reason: no more authentication methods available
2011-08-25 21:44:42+0200 [SSHClientTransport,client] connection lost
2011-08-25 21:44:42+0200 [SSHClientTransport,client] Stopping factory <twisted.conch.client.direct.SSHClientFactory instance at 0x244f320>
2011-08-25 21:44:42+0200 [-] Main loop terminated.
Connection to review.source.android.com closed.
2011-08-25 21:44:42+0200 [-] conch: exiting with error [Failure instance: Traceback (failure with no frames): <class 'twisted.conch.error.ConchError'>: ('no more authentication methods available', 14)
2011-08-25 21:44:42+0200 [-] ]

Attachments:

* [conch-bool.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e2/1e27b17392dabae2f5ba2a767b36d2dfc8a17275/07d40ba2e0525a3c2f9ed81b5c5d151938cac5d2.diff) (968 bytes) - added by toidinamai on 2011-08-25 21:52:20Z - 
* [conch-bool+test.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e2/1e27b17392dabae2f5ba2a767b36d2dfc8a17275/abe1d5b7da4ccc43c5df7c76b5d2db6db854cb27.diff) (1980 bytes) - added by toidinamai on 2011-08-25 21:59:41Z - Send \x01 instead of \xff as boolean TRUE when generating the publickey auth signature.

<details><summary>Searchable metadata</summary>

```
trac-id__5241 5241
type__defect defect
reporter__toidinamai toidinamai
priority__normal normal
milestone__ 
branch__branches_conch_publickey_bool_5241 branches/conch-publickey-bool-5241
branch_author__toidinamai toidinamai
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1314302059000000 1314302059000000
changetime__1314368363000000 1314368363000000
version__None None
owner__exarkun exarkun
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Step 4, "Try same with Conch" is pretty vague.  What does that mean?  Preferably give a command line.

**Comment by twisted-trac** [comment]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|toidinamai commented|
|-|-|

ssh -p 29418 -i ~/.ssh/keys/review.source.android.com.key -l frank.benkstein review.source.android.com

./bin/conch/conch -v -p 29418 -i ~/.ssh/keys/review.source.android.com.key -l frank.benkstein review.source.android.com

As far as I can tell the server accepts the public key in the first round
but doesn't accept the signed data.

**Comment by twisted-trac** [comment]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|toidinamai commented|
|-|-|

After consulting RFCs and the apache mina source code
http://mina.apache.org/sshd/source.html which is apparently the ssh server used
within gerrit, I think there is a bug in both:  conch sends wrong data and mina
is not very robust in the data it accepts.  RFC 4251 says:

  boolean
      A boolean value is stored as a single byte.  The value 0
      represents FALSE, and the value 1 represents TRUE.  All non-zero
      values MUST be interpreted as TRUE; however, applications MUST NOT
      store values other than 0 and 1.

When signing the data conch sends \xff instead of "\x01" which is what leads
to the signature validation error in mina.

The attached patch fixes the issue for me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Thanks!  Remember to add the _review_ keyword to make sure someone looks at the patches. :)  Also, [so you get your points.](http://twistedmatrix.com/highscores/) ;)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [32469]) Branching to 'conch-publickey-bool-5241'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [32470]) Apply conch-bool+test.diff 

refs [#5241](https://github.com/twisted/twisted/issues/5241)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun|
|-|-|

The fix and tests look fine.  The [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-publickey-bool-5241) look good.  I'll go ahead and merge it.  Thanks a lot.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [32472]) Merge conch-publickey-bool-5241

Author: toidinamai
Reviewer: exarkun
Fixes: [#5241](https://github.com/twisted/twisted/issues/5241)

Change the Conch userauth client to send \x01 instead of \xff when it is constructing
"publickey" messages.  This improves RFC compliance by following the definition of
the boolean type more closely.  It also fixes an interoperability issue between
Conch and Apache SSHD.
